### ISSUE TITLE
`cond(custom_vjp)` errors in `xmap`

### ISSUE THREAD

**Issue by sharadmv** [google/jax#9540]
Seems like the `xmap` axis substitution pass is failing when you nest a `custom_vjp` (and `custom_jvp` presumably) inside of a `cond`.

Repro:
```python
import jax
from jax.experimental import maps
import jax.numpy as jnp
from jax import lax

@jax.custom_vjp
def identity(x):
  return x
# Define dummy fwd/bwd (they are never called)
identity.defvjp(fwd=lambda: None, bwd=lambda: None)

def run(x):
  def true_fn(x):
    return identity(x)
  def false_fn(x):
    return x
  return lax.cond(True, true_fn, false_fn, x)
  
sharded_run = maps.xmap(run, in_axes=['batch', ...], out_axes=['batch', ...])
sharded_run(jnp.arange(8.))
```

@apaszke

**Comment by rajasekharporeddy** [comment]
Hi @sharadmv 

Looks like this issue has been resolved. I executed the mentioned code with JAX version 0.4.23 (CPU and GPU) and 0.3.25 (TPU). It executed without any error.
```python
import jax
from jax.experimental import maps
import jax.numpy as jnp
from jax import lax

@jax.custom_vjp
def identity(x):
  return x
# Define dummy fwd/bwd (they are never called)
identity.defvjp(fwd=lambda: None, bwd=lambda: None)

def run(x):
  def true_fn(x):
    return identity(x)
  def false_fn(x):
    return x
  return lax.cond(True, true_fn, false_fn, x)

sharded_run = maps.xmap(run, in_axes=['batch', ...], out_axes=['batch', ...])
sharded_run(jnp.arange(8.))
```
Output:
```
Array([0., 1., 2., 3., 4., 5., 6., 7.], dtype=float32)
```
Kindly find the gists on [CPU](https://colab.research.google.com/gist/rajasekharporeddy/6764fa29310665a65b31fe53608561e6/cpu_-9540.ipynb), [GPU](https://colab.research.google.com/gist/rajasekharporeddy/85fe1dc311cc378c1f659409643364a2/gpu_-9540.ipynb), and [TPU](https://colab.research.google.com/gist/rajasekharporeddy/14d6f64d7c5bc5424644969dd8d5c7b7/tpu_-9540.ipynb) for reference.

Thank you.

**Comment by mattjj** [comment]
Awesome, thanks for letting us know!
